商品購入機能

ken-forest opened the pull request:
# What
ログインしているユーザーが商品を買うための機能

# Why
商品を購入した記録を保存させるため。もう１つはクレジット決済機能を作って、
商品売買によってお金のやり取りを実現させるため。


＜Gyazoで撮影したもの＞

・必要な情報を適切に入力して「購入」ボタンを押すと、商品の購入ができる
※ユーザmoriがkenの出品した「Rubyの参考書」¥4000を購入
https://gyazo.com/c97b17f24f7e292c20a28dbc1eb8b44c
↓
※Soldoutの文字が出ている
・ログイン状態の場合でも、売却済みの商品には「購入画面に進む」ボタンが表示されない
https://gyazo.com/5be4d8a6b8d1019debe7148ddbf7ebc5


・入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される
https://gyazo.com/8df8790514b29e4481eead776c2cf1fd



・ログイン状態の場合でも、URLを直接入力して、自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する
①ユーザーkenがmoriの「イラストA」を購入(購入画面のURL)
https://gyazo.com/0f8ba72eec008d4d903e67f3ca7d2456
②再びユーザーkenが「イラストA」の購入画面に行くために直接URLを打ち込む
https://gyazo.com/2a4a7e2f8062b8345dc546321442cf3c


・ログイン状態の場合でも、URLを直接入力して自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移する
※ユーザーmoriが出品した「イラストA」の購入ページURL
https://gyazo.com/6dc71b8e182b7bb8a824d4b3efa88e0a
※そこへ侵入しようとするmori
https://gyazo.com/bab9e50044998d2ffac5b5ac34fd2756

↓※さらにそこへ侵入しようとするログインしていないユーザー
・ログアウト状態の場合は、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する
https://gyazo.com/df50f0b9c07f9b0478828f175b73cff6


・売却済みの商品は、画像上に「sold out」の文字が表示される(トップページと詳細ページ)
※上記でユーザーkenが購入した「イラストA」
https://gyazo.com/59ba5e066894ff1a68dd2620900c7fea

・ログイン状態の場合でも、売却済みの商品には「商品の編集」「削除」ボタンが表示されない(もう１つの「購入画面に進む」ボタンについては最初から２つ目の項目参照)
※ユーザーmori視点、購入された「イラストA」のページ
https://gyazo.com/fac23d8e778e8e8f7b5ef3a990cf01ea




〜追加のプルリクエスト〜
・ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する
①ユーザーmoriの「イラストB」の編集ページのURL
https://gyazo.com/cf467e25bb1c6d2afb84f4959bbad98d

②ユーザーkenが「イラストB」を購入
https://gyazo.com/b75cc6e652acaa9c741bd9f08a9fb8fc

③ユーザーmoriが「イラストB」の編集ページへ直接飛ぼうとする
https://gyazo.com/25d7e28c621d48112e497698a42da3c7


・テストコード
https://gyazo.com/24a85fff16be909751a078d5cb1b60b2

ken-forest committed:
フリマアプリ素材と名前を合わせるための変更

〜変更点〜
・全てのファイルの
purchase_record→order
・全てのファイルの
purchase_records→orders
に変更

ken-forest committed:
orderモデル・shipping_addressモデルの２つを作成

ken-forest committed:
modelsディレクトリ下にorder_shipping_address.rbファイルを作成

扱うカラム名を追加

ken-forest committed:
order_shipping_addressモデルの各カラムにバリデーションを記述

ken-forest committed:
order_shipping_addressモデルにデータを保存する記述を追加、その他

バリデーションの「validates :order_id」を削除

ken-forest committed:
ordersコントローラを作成

ken-forest committed:
ネスト構造を使い、ordersコントローラへのルーティングを記述

ken-forest committed:
orderモデル・shipping_addressモデルにアソシエーションを記述、その他コメントアウトしていたアソシエーションの記述も解放

ken-forest committed:
ordersコントローラのindexアクションに@order_shipping_addressを定義

ken-forest committed:
詳細ページの購入ボタンのパスを記述して、購入ページへ飛べるようにした

link_toのパスに
item_orders_path(@item.id)
を記述

ken-forest committed:
購入ページのform_withの記述を変更

もともとあった記述に、、
model: @order_shipping_address, url:  item_orders_path
、、を追加

ken-forest committed:
購入画面上部に、購入する商品情報が出るようにした

ken-forest committed:
app/views/orders/index.html.erbの一部のhogeを変更

購入画面の配送先住所を入力する箇所の該当ランにあるhogeを、書きなした

ken-forest committed:
ordersコントローラのcreateアクションの中身を記述

ken-forest committed:
クレジットカード入力ランのhogeを書き直した

ken-forest committed:
スペルミスの２箇所とストロングパラメータを修正

・app/views/orders/index.html.erbの
92行目のprefectureのスペルミスを修正
・order_shipping_addressモデルの
saveメソッド内、ShippingAddressがクラス名になっていなかったので修正
・ordersコントローラ内のストロングパラメータにて、ハッシュ統合部分に、、
item_id: params[:item_id]
、、を追加

ken-forest committed:
購入失敗時に、エラーメッセージが表示されるようにした

ken-forest committed:
RSpecGemを使い、購入成功の正常形テストコードを作成

ken-forest committed:
郵便番号が空では購入できない、異常形テストコードを追加

ken-forest committed:
新たに４つの異常形テストコードを追加

以下のテストコードを追加
・郵便番号に記号のハイフンが無いと購入できない
・郵便番号が全角文字が入っいると購入できない
・郵便番号が全角数字だと購入できない
・都道府県が選ばれていないと購入できない

ken-forest committed:
RspecのMySQLエラーを防ぐための記述を追加

beforeの箇所に、、
sleep 0.1
、、を追加。
これにより、Rspecのエラーが出なくなった。

ken-forest committed:
新たに異常形テストを多く追加

ken-forest committed:
クレジットカード関連の実装、card.jsが読み込まれるようにした

ken-forest committed:
公開鍵を設定し、フォームの情報を取得して、それをトークン化

ken-forest committed:
トークン化された値をフォームに埋め込み、それは非表示になるようにした

ken-forest committed:
カード情報をサーバに送らないように、フォームからカード情報を削除

ken-forest committed:
サーバでトークンを受け取れるようにした。

ken-forest committed:
PayjpGemを導入し、ordersコントローラのcreateアクションに決済処理を記述

ken-forest committed:
決済処理をpay_itemメソッドとして定義して、praivate以下のランへ移動させた

ken-forest committed:
order_shipping_addressモデルに、トークンのバリデーションを追加、その他

RSpecの購入記録のFactoryBotに、トークンの項目を追加

ken-forest committed:
トークンに関するテストコードを追加

・正常形を、トークンが存在すれば、、という条件を付け足した
・異常形に、トークンが空では購入できない、、を追加

ken-forest committed:
異常形テストに同じものがあったので削除

ken-forest committed:
トップページの商品一覧で、売れていればSoldOutが表示されるようにした

ken-forest committed:
詳細ページにて、売れていればSoldOutを表示するようにした

ken-forest committed:
ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されないようにした

ken-forest committed:
実装案内を削除。ローカルの環境変数を設定

ken-forest committed:
サーバーサイドで環境変数を読み込めるようにした

ordersコントローラのpay_itemメソッド、、PAY.JPテスト秘密鍵の箇所を環境変数に置き換えた

ken-forest committed:
クライアントサイドで環境変数を読み込めるように、JavaScriptで環境変数を呼び込めるようにした

ken-forest committed:
詳細ページにて、ログイン中なら編集・削除など各ボタンの出現の有無の箇所を少し変更

ken-forest committed:
ordersコントローラに記述追加

・authenticate_user!メソッドでログインしていないユーザーを弾くようにした
・redirect_to_rootメソッドを定義
これにより、自身が出品した商品の購入ページへ行けなくした。また商品が購入されている場合、購入ページへ行けなくした

ken-forest committed:
おかしな点を修正

app/views/items/edit.html.erbに遷移できなかったので、image_tagという記述を追加して飛べるようにした

ken-forest committed:
itemsコントローラのeditアクションに記述を追加

以下を追加、、
if @item.order.present?
      redirect_to root_path
 end
、、これにより、その商品が売却済みだと、編集ページに直接飛んで来れなくなる

ken-forest commented:
レビューをよろしくお願いします！

Kentozzz commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,62 @@
+
+  def index
+    @order_shipping_address = OrderShippingAddress.new
+    @item = Item.find(params[:item_id])
```
```
@item = Item.find(params[:item_id])
```
private以下でメソッド化しているため、こちらの記述は重複した記述となります！削除しておきましょう！13行目もredirect_to_rootを使用すれば記述不要となります。

Kentozzz commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,62 @@
+    
+    if @item.order.present?
+      redirect_to root_path
+    end
```
こちらの条件分岐は、以下の点から1行にまとめることができます。
```
①どちらもtrueの場合の記述のみなこと
②どちらもredirect_to root_pathと、遷移先が同じこと
```
2つのif文を一つにまとめてみましょう！

Kentozzz commented on app/javascript/card.js:
```diff
@@ -0,0 +1,38 @@
+
+    console.log(card);
+    Payjp.createToken(card, (status, response) => {
+      console.log(response);
```
```
console.log(response);
```
不要なconsole.logは削除しておきましょう！

Kentozzz commented on app/models/order_shipping_address.rb:
```diff
@@ -0,0 +1,25 @@
+
+
+  def save
+    # 購入情報を保存し、変数orderに代入する
```
```
# 購入情報を保存し、変数orderに代入する
```
コメントアウトは基本的に残さないようにしましょう！
もし残す場合は、なぜ残す必要があるかなどを記載いただけると幸いです！

Kentozzz commented on spec/models/order_shipping_address_spec.rb:
```diff
@@ -0,0 +1,107 @@
+      it 'postal_code、prefecture_id、city、address、phone_numberが正しく入力されている。かつ、トークンが存在すれば購入できる' do
+        expect(@order_shipping_address).to be_valid
+      end
+
```
以下のテストの実行も追加をお願いします！
```
建物名が空でも保存できること
```
建物名が空でも保存できることをテストするために、最初にFactoryBotで生成される建物名が常に入っている状態を作る必要がございます！まずはFactoryBotの修正からおこないましょう！

Kentozzz commented on spec/models/order_shipping_address_spec.rb:
```diff
@@ -0,0 +1,107 @@
+        @order_shipping_address.valid?
+        expect(@order_shipping_address.errors.full_messages).to include("Token can't be blank")
+      end
+
```
以下のテストも実施お願いします！
```
電話番号が9桁以下では保存できないこと
電話番号が12桁以上では保存できないこと
userが紐付いていなければ保存できないこと
itemが紐付いていなければ保存できないこと
```

Kentozzz reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認をお願いいたします！

ken-forest committed:
コメントアウトを削除

ken-forest committed:
console.logを削除

ken-forest committed:
テストコードを追加、その他

・order_shipping_addressesのFactoryBotに建物情報を追加で定義
・正常形テストコードを建物情報がある場合とない場合の２つに分けた

ken-forest committed:
ordersコントローラの記述を変更

「@item = Item.find(params[:item_id])」を定義してある、redirect_to_rootメソッドは、indexとcreateアクションのbefore_actionとして発動するようにしたので、２つのアクション内から
「@item = Item.find(params[:item_id])」を削除

ken-forest committed:
ordersコントローラのredirect_to_rootメソッドの記述を変更

・if current_user == @item.user
・if @item.order.present?
この２つの条件式を「&&」を使い、１つの条件式にまとめた

ken-forest committed:
異常形テストコードを２つ追加

・ユーザー情報が紐づいていないと購入できない
・購入したい商品の情報が紐づいていないと購入できない
、、以上の２つのテストコードを追加

ken-forest committed:
購入画面で入力する電話番号のバリデーションを変更

条件を「半角数字のみ」から「最低10ケタ、最高11ケタの0~9の半角数字のみ」に変更

ken-forest committed:
新たに２つの異常系テストコードを追加

・電話番号が9ケタ以下だったら購入できない
・電話番号が12ケタ以上だったら購入できない
、、以上の２つのテストコードを追加

ken-forest commented:
ご指摘いただいた箇所を修正させていただきました。
引き続き、レビューをよろしくお願いします！

＜Gyazoで撮影した修正後のテストコード＞
https://gyazo.com/80b1f4d38849a4e536a79c561aa12da5

Kentozzz commented on app/views/items/show.html.erb:
```diff
@@ -23,19 +25,25 @@
 
         <% if current_user.id == @item.user_id %>
 
+              <%# <% unless @item.order.present? %>
```
```
<%# <% unless @item.order.present? %>
```
こちらのコメントアウトは不要かと存じますので、削除をお願いします！

Kentozzz commented on app/views/items/show.html.erb:
```diff
@@ -23,19 +25,25 @@
               <%= link_to "商品の編集", edit_item_path(@item.id), method: :get, class: "item-red-btn" %>
               <p class="or-text">or</p>
               <%= link_to "削除", item_path(@item.id), method: :delete, class:"item-destroy" %>
+              <%# <% end %>
```
```
<%# <% end %>
```
こちらも同様に削除をお願いします！

Kentozzz commented on spec/models/order_shipping_address_spec.rb:
```diff
@@ -0,0 +1,136 @@
+      end
+
+      it '電話番号が9ケタ以下だったら購入できない' do
+        @order_shipping_address.phone_number = '07436633'
```
```
@order_shipping_address.phone_number = '07436633'
```
非常に細かい点となりますが、代入された桁数が8桁となっておりますので、9桁に修正しましょう。
※9桁でテストをすることにより正確性が高まります。

Kentozzz reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
非常に細かい点とはなりますが、追加でコメントいたしましたのでご確認お願いします！
あと少しです！応援しております！

ken-forest committed:
コメントアウトの削除、異常系テストコードの記述変更

商品の詳細ページにて、
・<%# <% unless @item.order.present? %>
・<%# <% end %>
、、以上の２つのコメントアウトを削除
・電話番号が9ケタ以下だったら購入できない異常系テストコードの、電話番号の記述を9ケタに変更

ken-forest commented:
細かい点まで見ていただき、ありがとうございます。
修正しましたので再々ですが、レビューをよろしくお願いします！

＜Gyazoで撮影した修正後のテストコード＞
https://gyazo.com/20af85ddaedbb1053c9263d1a1c4c165

Kentozzz reviewed (commented):
レビュー依頼ありがとうございます！
LGTMとさせていただきます！実装お疲れ様でした！
このブランチをマージし、masterブランチでプルを行いましょう！
機能が反映されていることを確認し、引き続き実装頑張っていきましょう！
